### PR TITLE
538 prerequisite bulk upload

### PR DESCRIPTION
**Pull request by bell-pa**
## Context

Trello: [Bulk upload of Prerequisites](https://trello.com/c/OTT2z5KI/538-endpoint-for-uploading-prerequisites)

This is the second of three pull requests which together will provide bulk up-load functionality of accredited programmes data originating from Google spreadsheets produced by the team and exported to CSV files.

The [first pull request](https://github.com/ministryofjustice/hmpps-accredited-programmes-api/pull/57) provides background on this work.

## Changes in this PR
This pull request adds a PUT /courses/prerequisites end-point.  The end-point accepts data in CSV format.
Each row of data represents one course prerequisite. Each row should contain:
* name: The name of this prerequisite. Not unique.
* course: The name of the course to which this prerequisite applies.
* description: The description of the prerequisite.
* comments: Any additional information

Neither name, description or course are independently unique. However the business key <name, course, description> should be unique.

On upload the application deletes any stored prerequisites before adding the new values. Courses themselves are untouched. Each prerequisite must identify the course to which it applies. That is, the `course` value in the prerequisite row must match the `name` of a course.  The prerequisite is then added to the course.  When a match can't be found the prerequisite is ignored.  

There is no feed back on the outcome of the upload.  Responses that may help with debugging may be added in future pull requests.

Please refer to the commit messages for information on the purpose and content of each commit.

### CODE REVIEW

**Comment by bell-pa** [comment]
Superseded by PR https://github.com/ministryofjustice/hmpps-accredited-programmes-api/pull/60 
That PR contains both the work in the PR and the work for bulk offerings upload.

**Review by Gweaton — COMMENTED** [review]
This looks good to me, thanks for the atomic commits and descriptive messages, it made reviewing this very easy.

Just a couple of minor things

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/domain/CourseServiceTest.kt [hunk 1]
```diff
@@ -153,4 +150,14 @@ class CourseServiceTest {
       ),
     )
   }
+
+  @Test
+  fun `replaceAllPrerequisites - course name mismatch - record ignored`() {
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/domain/CourseServiceTest.kt:155]: Ah, I can see it's in the other PR.

**Comment by bell-pa** [comment]
Superseded by PR 60 and therefore withdrawn

**Pull request closed without merging**